frontend/docsettings.lua

```diff
@@ -130,63 +139,58 @@ function DocSettings:getLastSaveTime(doc_path)
     end
 end
 
-function DocSettings:ensureSidecar(sidecar)
-    if lfs.attributes(sidecar, "mode") ~= "directory" then
-        lfs.mkdir(sidecar)
-    end
-end
-
 --- Opens a document's individual settings (font, margin, dictionary, etc.)
--- @string docfile path to the document (e.g., `/foo/bar.pdf`)
+-- @string doc_path path to the document (e.g., `/foo/bar.pdf`)
 -- @treturn DocSettings object
-function DocSettings:open(docfile)
-    --- @todo (zijiehe): Remove history_path, use only sidecar.
-
+function DocSettings:open(doc_path)
     -- NOTE: Beware, our new instance is new, but self is still DocSettings!
     local new = DocSettings:extend{}
-    new.history_file = new:getHistoryPath(docfile)
 
-    local sidecar = new:getSidecarDir(docfile)
-    new.sidecar = sidecar
-    DocSettings:ensureSidecar(sidecar)
-    -- If there is a file which has a same name as the sidecar directory,
-    -- or the file system is read-only, we should not waste time to read it.
-    if lfs.attributes(sidecar, "mode") == "directory" then
-        -- New sidecar file name is metadata.{file last suffix}.lua.
-        -- So we can handle two files with only different suffixes.
-        new.sidecar_file = new:getSidecarFile(docfile)
-        new.legacy_sidecar_file = sidecar.."/"..
-                                  ffiutil.basename(docfile)..".lua"
+    new.sidecar_dir_doc = new:getSidecarDir(doc_path, ":doc")
+    new.sidecar_file_doc = new:getSidecarFile(doc_path, ":doc")
+    local sidecar_file_doc, legacy_sidecar_file
+    if lfs.attributes(new.sidecar_dir_doc, "mode") == "directory" then
+        sidecar_file_doc = new.sidecar_file_doc
+        legacy_sidecar_file = new.sidecar_dir_doc.."/"..ffiutil.basename(doc_path)..".lua"
     end
+    new.sidecar_dir_sidecars = new:getSidecarDir(doc_path, ":sidecars")
+    new.sidecar_file_sidecars = new:getSidecarFile(doc_path, ":sidecars")
+    local sidecar_file_sidecars
+    if lfs.attributes(new.sidecar_dir_sidecars, "mode") == "directory" then
+        sidecar_file_sidecars = new.sidecar_file_sidecars
+    end
+    local history_file = new:getHistoryPath(doc_path)
 
     -- Candidates list, in order of priority:
     local candidates_list = {
-        -- New sidecar file
-        new.sidecar_file or "",
-        -- Backup file of new sidecar file
-        new.sidecar_file and (new.sidecar_file .. ".old") or "",
+        -- New sidecar file in doc folder
+        sidecar_file_doc or "",
+        -- Backup file of new sidecar file in doc folder
+        sidecar_file_doc and (sidecar_file_doc..".old") or "",
         -- Legacy sidecar file
-        new.legacy_sidecar_file or "",
+        legacy_sidecar_file or "",
+        -- New sidecar file in sidecars folder
+        sidecar_file_sidecars or "",
+        -- Backup file of new sidecar file in sidecars folder
+        sidecar_file_sidecars and (sidecar_file_sidecars..".old") or "",
```

Comment:
I guess the ordering here doesn't matter and dont enforce a priority - and that it's the mtime of the files found that will set the priority order, right?

Reply:
The priority is *always* enforced; as a practical example, if you were to migrate from sidecar to centralized sidecar while preserving mtime (and not touching the document so as to affect said sidecar data), the doc's sidecar would be preferred because they have higher priority.

(Which leads to an obvious: if you do intend to create a migration path, the previous location should be cleared, or the new location files "touched" so as to get an updated mtime).

Reply:
We'd better have `sidecar.lua` before `sidecar.lua.old`
https://github.com/koreader/koreader/blob/50b2267b74e94a4831213a4caa6cde64dbe76eb8/frontend/docsettings.lua#L26-L39
And finally the candidates are sorted by `mtime`, yes.

Reply:
I didn't think about the migration of sidecar files for **all books**. Is it needed?
The migration for a certain book is done by opening it.

Reply:
Yes, please, no migration ! :)
We don't know all possible book locations.
And there's always the case of removable SD cards, that may not be plugged in at the time when the migration is done.

> The priority is always enforced; as a practical example, if you were to migrate from sidecar to centralized sidecar while preserving mtime and leaving the old sidecar files in place (and not touching the document so as to affect said sidecar data), the doc's sidecar would be preferred because they have higher priority (that rule only applies if mtime matches, otherwise the primary sort order is by MRU).

~~So, I guess this is bad ? :/~~
~~The case we want to handle is:~~
- ~~we have book A and book B with "alongside doc" sidecars.~~
- ~~user changes setting to "centralized" (no migration is done)~~
- ~~user opens book A: "no centralized sidecar" found, "alongside doc" is found, use that~~
- ~~user closes book A: "centralized sidecar" is saved (should "alongside doc" be deleted?)~~
- ~~user changes setting to "alongside docs" (no migration is done)~~
- ~~user opens book A: "alongside doc" is found and used first, but "centralized sidecar" is newer and should have been used~~

Scratch that, if you really mean MRU (Most Recently User), which is what I wanted :)
(that is, the order the sidecar candidates are listed in the code does not matter - we don't copy preserving mtime, so there will always be a newer, that will be used).

Reply:
> the order the sidecar candidates are listed in the code does not matter

It does (in a technical sense, at least), but it's only used as a tie-breaker if there's an mtime conflict.

If there's no mtime conflict, mtime wins ;).